### PR TITLE
[ESM] Add backoff between Stream Poller retries

### PR DESCRIPTION
**Pull request by gregfurman**
## Motivation
This PR leverages the backoff util to ensure, when retrying, we wait between retries so as to not spam the processing loop.

Companion PR to https://github.com/localstack/localstack/pull/12264

## Changes
- Checks `has_record_expired` every iteration to allow for an expired record to break processing.
- Use `backoff`utility, in conjunction with a `threading.Event`, to exponentially back-off between processing retries.

### CODE REVIEW

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py [hunk 1]
```diff
@@ -187,9 +210,26 @@ def poll_events_from_shard(self, shard_id: str, shard_iterator: str):
...
+                    abort_condition = "RecordAgeExpired"
+
+                if attempts > 0:
+                    # TODO: Should we always backoff (with jitter) before processing since we may not want multiple pollers
+                    # all starting up and polling simultaneously
+                    # For example: 500 persisted ESMs starting up and requesting concurrently could flood gateway
+                    self._is_shutdown.wait(boff.next_backoff())
+
                 self.processor.process_events_batch(events)
+                boff.reset()
```
Comment (gregfurman) [inline localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py:231]: @tiurin An example of how we can use the new backoff utility 👆👀

**Review by tiurin — APPROVED** [review]

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py [hunk 2]
```diff
@@ -187,9 +210,26 @@ def poll_events_from_shard(self, shard_id: str, shard_iterator: str):
...
+            not abort_condition
+            and not self.max_retries_exceeded(attempts)
+            and not self._is_shutdown.is_set()
+        ):
             try:
+                if self.has_record_expired(polled_events[-1]):
+                    abort_condition = "RecordAgeExpired"
+
+                if attempts > 0:
+                    # TODO: Should we always backoff (with jitter) before processing since we may not want multiple pollers
```
Comment (joe4dev) [inline localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py:225]: Starting with backoff after a retry seems a good solution to start with for now, not impacting any best-case responsiveness

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py [hunk 3]
```diff
@@ -104,12 +111,29 @@ def format_datetime(self, time: datetime) -> str:
...
         return events
 
+    def has_record_expired(self, event: dict):
+        # Check MaximumRecordAgeInSeconds
+        if maximum_record_age_in_seconds := self.stream_parameters.get("MaximumRecordAgeInSeconds"):
+            arrival_timestamp_of_last_event = event.get("approximateArrivalTimestamp")
+            if not arrival_timestamp_of_last_event:
+                return False
+
+            now = get_current_time().timestamp()
```
Comment (joe4dev) [inline localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py:130]: Are we handling timezones correctly here?

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py [hunk 4]
```diff
@@ -104,12 +111,29 @@ def format_datetime(self, time: datetime) -> str:
...
+    def close(self):
+        self._is_shutdown.set()
+
     def pre_filter(self, events: list[dict]) -> list[dict]:
         return events
 
     def post_filter(self, events: list[dict]) -> list[dict]:
         return events
 
+    def has_record_expired(self, event: dict):
```
Comment (joe4dev) [inline localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py:123]: That's an untested feature for which we should ideally add a test (which is a bit annoying because of the 60s minimum https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventsourcemapping.html#cfn-lambda-eventsourcemapping-maximumrecordageinseconds). Can we at least add a backlog item for this?

**Review by joe4dev — APPROVED** [review]
Nice application of the backoff utility 👍 

It would be great to highlight the motivation on why we are doing this from a customer perspective (as discussed and raised by @dfangl ) in future PR descriptions.

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py (hunk 4 above)
Comment (gregfurman) [reply localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py:123]: I'm happy to revert these changes and make a follow-up PR with a test instead?

Comment (dfangl) [reply localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py:123]: This is blocking for one of my PRs (as the logs are so massive they cannot reasonably be printed due to the retries), would be nice to get this in a separate PR, as there seem to be some confusion still.

Comment (gregfurman) [reply localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py]: Done!

**Review by dfangl — APPROVED** [review]
Looks good, thanks for the fix! No more 200k log lines after the function was deleted :)
